dbt/adapters/snowflake/relation_configs/dynamic_table.py

```diff
@@ -21,6 +35,147 @@ class SnowflakeDynamicTableConfig(RelationConfigBase):
+        return dynamic_table
+
+    @classmethod
+    def parse_model_node(cls, model_node: ModelNode) -> dict:
+        config_dict = {
+            "name": model_node.identifier,
+            "schema_name": model_node.schema,
+            "database_name": model_node.database,
+            "query": model_node.compiled_code,
+            "target_lag": SnowflakeDynamicTableTargetLagConfig.parse_model_node(model_node),
+            "warehouse": model_node.config.extra.get("warehouse"),
+        }
+        return config_dict
+
+    @classmethod
+    def parse_relation_results(cls, relation_results: RelationResults) -> dict:
+        if dynamic_table := relation_results.get("dynamic_table"):
+            dynamic_table_config = dynamic_table.rows[0]
+        else:
+            dynamic_table_config = {}
+
+        config_dict = {
+            "name": dynamic_table_config.get("name"),
+            "schema_name": dynamic_table_config.get("schema_name"),
+            "database_name": dynamic_table_config.get("database_name"),
+            "query": cls._parse_query(dynamic_table_config.get("text")),
+            "target_lag": SnowflakeDynamicTableTargetLagConfig.parse_relation_results(
+                relation_results
+            ),
+            "warehouse": dynamic_table_config.get("warehouse"),
+        }
+        return config_dict
+
+    @classmethod
+    def _parse_query(cls, query: str) -> str:
+        """
+        Get the select statement from the dynamic table definition in Snowflake.
+
+        Args:
+            query: the `create dynamic table` statement from `show dynamic tables`, for example:
+
+            create dynamic table my_dynamic_table
+                target_lag = '1 minute'
+                warehouse = MY_WAREHOUSE
+                as (
+                    select * from my_base_table
+                )
+            ;
+
+        Returns: the `select ...` statement, for example:
+
+            select * from my_base_table
+
+        """
+        open_paren = query.find("as (") + len("as (")
+        close_paren = query.rindex(")")
+        return query[open_paren:close_paren].strip()
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class SnowflakeDynamicTableWarehouseConfigChange(
+    RelationConfigChange, RelationConfigValidationMixin
+):
+    context: str
+
+    @property
+    def requires_full_refresh(self) -> bool:
+        return False
+
+    @property
+    def validation_rules(self) -> Set[RelationConfigValidationRule]:
+        return {
+            RelationConfigValidationRule(
+                validation_check=self.action == RelationConfigChangeAction.alter,
+                validation_error=DbtRuntimeError(
+                    f"Warehouse should only be altered but {self.action} was received."
+                ),
+            ),
+        }
+
+
+@dataclass
+class SnowflakeDynamicTableConfigChangeset:
+    target_lag: Optional[SnowflakeDynamicTableTargetLagConfigChange] = None
```

Comment:
Can't remember. Could you just have the vars without the `= None` and get the same thing? Are you just erring to explicit? Is one considered cleaner? Can I find a way to ask more questions about this one line?

Reply:
If I remember correctly, `dataclass` yells at you if you have an `Optional` and you don't provide the variable because you haven't defined a default, even if that default is `None`. I think it doesn't assume. That being said, these are one of two varieties. There can either be a change that can only happen once (e.g. `target_lag`), or a set of changes all of the same type (e.g. `indexes` on Postgres). So I'm also being explicit that the null case here is `None` and not `set()`. I've also thought about collapsing this entirely and just having one single set of changes, but that leaves the parsing of "what _type_ of change is this?" to the caller, which in this case is jinja.